internal/storage/sort.go

```diff
@@ -172,21 +172,40 @@ func NewPriorityQueue[T any](less func(x, y *T) bool) *PriorityQueue[T] {
 func MergeSort(batchSize uint64, schema *schemapb.CollectionSchema, rr []RecordReader,
 	rw RecordWriter, predicate func(r Record, ri, i int) bool,
 ) (numRows int, err error) {
+	// Fast path: no readers provided
+	if len(rr) == 0 {
+		return 0, nil
+	}
+
 	type index struct {
 		ri int
 		i  int
 	}
 
 	recs := make([]Record, len(rr))
+	// advanceRecord reads next record for reader i while releasing the previous one to
+	// avoid memory leaks introduced by cgo Arrow arrays.
 	advanceRecord := func(i int) error {
-		rec, err := rr[i].Next()
-		recs[i] = rec // assign nil if err
-		if err != nil {
-			return err
+		// Release previous record if any.
+		if recs[i] != nil {
+			recs[i].Release()
+			recs[i] = nil
 		}
-		return nil
+
+		rec, err := rr[i].Next()
+		recs[i] = rec
+		return err
 	}
 
+	// Ensure all remaining records are released on return.
+	defer func() {
+		for _, r := range recs {
+			if r != nil {
+				r.Release()
+			}
+		}
+	}()
```

Comment:
releasing here is not needed
The design rule here is that: the ownership belongs to reader here. as long as the reader is closed, all the record from it shall be release properly.